### PR TITLE
Provide a custom merge operator for MVCCStats

### PR DESCRIPTION
**Pull request by spencerkimball**
Also added two new stats to track for system keys. The range data iterator is now used with MVCCComputeStats.

With custom merge iterator, merge performance seems greatly improved except when reading once every 1000 merge writes. I don't really believe that it should get slower vs. doing 12x the number of single integer value merges.

### CODE REVIEW

**Comment by spencerkimball** [comment]
`benchmark                           old ns/op     new ns/op     delta
BenchmarkMVCCStatsMerge1For1        9840579       7039319       -28.47%
BenchmarkMVCCStatsMerge10For1       2105464       719971        -65.80%
BenchmarkMVCCStatsMerge100For1      239554        145619        -39.21%
BenchmarkMVCCStatsMerge1000For1     103913        149536        +43.90%

benchmark                           old MB/s     new MB/s     speedup
BenchmarkMVCCStatsMerge1For1        0.01         0.02         2.00x
BenchmarkMVCCStatsMerge10For1       0.05         0.18         3.60x
BenchmarkMVCCStatsMerge100For1      0.47         0.88         1.87x
BenchmarkMVCCStatsMerge1000For1     1.08         0.86         0.80x
`

File: proto/data.go [hunk 1]
```diff
@@ -636,6 +636,8 @@ func (ms *MVCCStats) Add(oms *MVCCStats) {
 	ms.IntentCount += oms.IntentCount
 	ms.IntentAge += oms.IntentAge
 	ms.GCBytesAge += oms.GCBytesAge
+	ms.SysBytes += oms.SysBytes
+	ms.SysCount += oms.SysCount
 	ms.LastUpdateNanos += oms.LastUpdateNanos
```
Comment (mrtracy) [inline proto/data.go]: I noticed this earlier; we're not treating LastUpdateNanos correctly, I don't think it's simply a matter of addition.

File: proto/data.go [hunk 2]
```diff
@@ -651,9 +653,39 @@ func (ms *MVCCStats) Subtract(oms *MVCCStats) {
 	ms.IntentCount -= oms.IntentCount
 	ms.IntentAge -= oms.IntentAge
 	ms.GCBytesAge -= oms.GCBytesAge
+	ms.SysBytes -= oms.SysBytes
+	ms.SysCount -= oms.SysCount
 	ms.LastUpdateNanos -= oms.LastUpdateNanos
 }
 
+// ToValue generates a Value message which contains an encoded copy of
+// the staats in its "bytes" field. The returned Value will also have
```
Comment (mrtracy) [inline proto/data.go]: s/staats/stats

**Comment by spencerkimball** [comment]
Eliminating the merge completely yielded dramatic speedup:

`  300000	      5122 ns/op	  24.99 MB/s`

**Comment by spencerkimball** [comment]
PTAL

**Comment by tschottdorf** [comment]
Wait, are you saying that if we don't use the merge operator at all we're fastest by a wide margin?

File: storage/engine/mvcc.go [hunk 3]
```diff
@@ -251,41 +209,51 @@ func updateStatsOnResolve(ms *proto.MVCCStats, key proto.Key, origMetaKeySize, o
...
+	ok, sys := updateStatsForKey(ms, key)
+	if !ok {
 		return
 	}
 	origTotalBytes := orig.KeyBytes + orig.ValBytes + origMetaKeySize + origMetaValSize
-	if !orig.Deleted {
-		ms.LiveBytes -= origTotalBytes
-		ms.LiveCount--
+	if sys {
+		ms.SysBytes -= origTotalBytes
```
Comment (petermattis) [inline storage/engine/mvcc.go]: Don't you need to condition on `!orig.Deleted` here?

Comment (spencerkimball) [inline storage/engine/mvcc.go]: No, for the system keys we're not differentiating between live and dead bytes. It's just a catch all for key count and bytes count stored in the system-local keyspace for each range.

File: proto/data.go (hunk 1 above)
Comment (spencerkimball) [inline proto/data.go]: We're currently treating the update nanos as a delta when we call this add method, so it's being used correctly.

**Comment by petermattis** [comment]
LGTM

**Pull request merged**